### PR TITLE
Symbol annotations on core

### PR DESCRIPTION
**Pull request by kohsuke**
To help [structs-plugin](https://wiki.jenkins-ci.org/display/JENKINS/Structs+plugin) and its users (such as pipeline, system-config-dsl, etc), we should have core classes annotated with `@Symbol`

Strictly speaking, this doesn't meet the criteria to be in the 2.0 branch right now, but this is such a useful yet safe (from the perspective of regressions) changes...

@reviewbybees esp. @daniel-beck as the release manager.

### CODE REVIEW

**Comment by jtnord** [comment]
So I am :bug on this for the moment for the following reasons.

1. it brings in a new bundled plugin at the time we are trying to remove as many as possible. (we should be able to use the annotations without bundling it by setting the correct scope and exclusions in the various projects)
2.   according to the metadata on the [plugin page](https://wiki.jenkins-ci.org/display/JENKINS/Structs+plugin) the plugin is in subversion and @rtyler is planning to just terminate that service - meaning we can not maintain it ;-)

Both of these should be trivial to fix, at which point I am happy to remove my :bug:

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 1]
```diff
@@ -1593,6 +1594,22 @@ public void addView(View v) throws IOException {
         viewGroupMixIn.addView(v);
     }
 
+    /**
+     * Completely replaces views
+     */
+    @DataBoundSetter
+    public void setViews(Collection<View> views) throws IOException {
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: Seems to be unrelated on the face of it?

File: core/src/main/java/hudson/diagnosis/TooManyJobsButNoView.java [hunk 2]
```diff
@@ -39,7 +40,7 @@
  *
  * @author Kohsuke Kawaguchi
  */
-@Extension
+@Extension @Symbol("tooManyJobsNoView")
```
Comment (jtnord) [inline core/src/main/java/hudson/diagnosis/TooManyJobsButNoView.java]: should this have been `tooManyJobsButNoView`?

File: core/src/main/java/hudson/model/ListView.java [hunk 3]
```diff
@@ -109,6 +111,14 @@ public ListView(String name, ViewGroup owner) {
         this.owner = owner;
     }
 
+    /**
+     * Sets the columns of this view.
+     */
+    @DataBoundSetter
```
Comment (jtnord) [inline core/src/main/java/hudson/model/ListView.java]: unrelated change?

File: core/src/main/java/hudson/model/Slave.java [hunk 4]
```diff
@@ -162,6 +163,16 @@ public Slave(String name, String nodeDescription, String remoteFS, int numExecut
     	this(name, nodeDescription, remoteFS, numExecutors, mode, labelString, launcher, retentionStrategy, new ArrayList());
     }
 
+    public Slave(@Nonnull String name, String remoteFS, ComputerLauncher launcher) throws FormException, IOException {
```
Comment (jtnord) [inline core/src/main/java/hudson/model/Slave.java]: unrelated changes in this file?

File: core/src/main/java/hudson/model/TextParameterDefinition.java [hunk 5]
```diff
@@ -37,7 +38,7 @@ public TextParameterDefinition(String name, String defaultValue, String descript
         super(name, defaultValue, description);
     }
 
-    @Extension
+    @Extension @Symbol({"text","textParam"})
```
Comment (jtnord) [inline core/src/main/java/hudson/model/TextParameterDefinition.java]: why do parameter types all get 2 definitions?  What is special about them?

File: core/src/main/java/hudson/node_monitors/ClockMonitor.java [hunk 6]
```diff
@@ -46,8 +49,19 @@ public ClockDifference getDifferenceFor(Computer c) {
         return DESCRIPTOR.get(c);
     }
 
-    @Extension
-    public static final AbstractNodeMonitorDescriptor<ClockDifference> DESCRIPTOR = new AbstractAsyncNodeMonitorDescriptor<ClockDifference>() {
+    /**
```
Comment (jtnord) [inline core/src/main/java/hudson/node_monitors/ClockMonitor.java]: this would seem to be subject for a different PR?

File: core/src/main/java/hudson/node_monitors/SwapSpaceMonitor.java [hunk 7]
```diff
@@ -80,8 +81,18 @@ public String getColumnCaption() {
         return Jenkins.getInstance().hasPermission(Jenkins.ADMINISTER) ? super.getColumnCaption() : null;
     }
 
-    @Extension
-    public static final AbstractNodeMonitorDescriptor<MemoryUsage> DESCRIPTOR = new AbstractAsyncNodeMonitorDescriptor<MemoryUsage>() {
+    /**
+     * @deprecated as of 2.0
```
Comment (jtnord) [inline core/src/main/java/hudson/node_monitors/SwapSpaceMonitor.java]: this would seem to be subject for a different PR?

File: core/src/main/java/hudson/node_monitors/TemporarySpaceMonitor.java [hunk 8]
```diff
@@ -61,7 +62,18 @@ public String getColumnCaption() {
         return Jenkins.getInstance().hasPermission(Jenkins.ADMINISTER) ? super.getColumnCaption() : null;
     }
 
-    public static final DiskSpaceMonitorDescriptor DESCRIPTOR = new DiskSpaceMonitorDescriptor() {
+    /**
+     * @deprecated as of 2.0
```
Comment (jtnord) [inline core/src/main/java/hudson/node_monitors/TemporarySpaceMonitor.java]: this would seem to be subject for a different PR?

File: core/src/main/java/hudson/security/FullControlOnceLoggedInAuthorizationStrategy.java [hunk 9]
```diff
@@ -92,7 +93,7 @@ public void setAllowAnonymousRead(boolean allowAnonymousRead) {
     @Restricted(NoExternalUse.class)
     public static Descriptor<AuthorizationStrategy> DESCRIPTOR;
 
-    @Extension
+    @Extension @Symbol("fullControlOnceLoggedIn")
```
Comment (jtnord) [inline core/src/main/java/hudson/security/FullControlOnceLoggedInAuthorizationStrategy.java]: `NullSCM` seemed to be matching the Display rather than the class (`none` vs `null`) - yet this is doing the opposite (`fullControlOnceLoggedIn` vs `loggedInUsersCanDoAnything`?

File: core/src/main/java/hudson/security/LegacySecurityRealm.java [hunk 10]
```diff
@@ -88,8 +91,19 @@ public Filter createFilter(FilterConfig filterConfig) {
         return (Filter) context.getBean("legacy");
     }
 
-    @Extension
-    public static final Descriptor<SecurityRealm> DESCRIPTOR = new Descriptor<SecurityRealm>() {
+    /**
```
Comment (jtnord) [inline core/src/main/java/hudson/security/LegacySecurityRealm.java]: seems to be unrelated to this PR?

File: core/src/main/java/hudson/tasks/BatchFile.java [hunk 11]
```diff
@@ -59,7 +60,7 @@ private Object readResolve() throws ObjectStreamException {
         return new BatchFile(command);
     }
 
-    @Extension
+    @Extension @Symbol("batchFile")
```
Comment (jtnord) [inline core/src/main/java/hudson/tasks/BatchFile.java]: this is "batchFile" vs shell is just "shell" not "shellScript"?

**Comment by jglick** [comment]
> it brings in a new bundled library plugin

No it does not. There is a _library_ and a _plugin_. The split was done specifically for the purpose of letting core depend only on the library.

> according to the metadata on the plugin page the plugin is in subversion

It is in GitHub and its Maven metadata [says so](https://github.com/jenkinsci/structs-plugin/blob/structs-parent-1.1/pom.xml#L21-L26). I have no idea why the Confluence macro is not getting this.

**Comment by jglick** [comment]
> this doesn't meet the criteria to be in the 2.0 branch right now

Why are you even proposing this for the 2.0 branch? I see no reason why this should not go into `master`.

File: core/src/main/java/hudson/AboutJenkins.java [hunk 12]
```diff
@@ -10,7 +12,7 @@
  * 
  * @author Kohsuke Kawaguchi
  */
-@Extension
+@Extension @Symbol("about")
```
Comment (jglick) [inline core/src/main/java/hudson/AboutJenkins.java]: It is pointless to be applying symbols to random extensions like this. We only care about `Descriptor`s. So a :bug: for polluting the codebase with symbols with no known use.

File: core/src/main/java/hudson/model/BooleanParameterDefinition.java [hunk 13]
```diff
@@ -72,7 +73,7 @@ public BooleanParameterValue getDefaultParameterValue() {
         return new BooleanParameterValue(getName(), defaultValue, getDescription());
     }
 
-    @Extension
+    @Extension @Symbol("booleanParam")
```
Comment (jglick) [inline core/src/main/java/hudson/model/BooleanParameterDefinition.java]: Other than possible reserved-word clashes, I would think this would just be `boolean` (similarly `choice`, `file`, etc.).

File: core/src/main/java/hudson/model/ListView.java (hunk 3 above)
Comment (jglick) [inline core/src/main/java/hudson/model/ListView.java]: I suppose this is intended to allow list views to be configured with a `columns` parameter from a script.

**Comment by jtnord** [comment]
>> it brings in a new bundled library plugin
> No it does not. There is a library and a plugin. The split was done specifically for the purpose of letting core depend only on the library.

Damn I though I had removed that plugin in the edit.
The plugin depends on this library - and there is no way of updating this library without updating jenkins core (as it is bundled - without doing a load of plugin first and making all plugins that want to use this ship with the library).  So if a newer plugin needs a newer library you are SOOL.  

The library AFAICT does not need to be shipped with Jenkins war - as annotations are designed to fail safe if the annotation is not on the classpath.  The way of accessing these annotations as I understand it is via the plugin - so if the plugin provided this library then it would be available to it and all other plugins that depended on it - so there should be no need to tie the annotation library version to a core version.

File: core/src/main/java/hudson/model/Slave.java (hunk 4 above)
Comment (jglick) [inline core/src/main/java/hudson/model/Slave.java]: For use from the new `DumbSlave` constructor, though :ant: this one should really be `protected` to clarify that.

File: core/src/main/java/hudson/model/TextParameterDefinition.java (hunk 5 above)
Comment (jglick) [inline core/src/main/java/hudson/model/TextParameterDefinition.java]: I was wondering the same.

File: core/src/main/java/hudson/model/StringParameterDefinition.java [hunk 14]
```diff
@@ -69,7 +70,7 @@ public StringParameterValue getDefaultParameterValue() {
         return v;
     }
 
-    @Extension
+    @Extension @Symbol({"string","stringParam"})
```
Comment (jglick) [inline core/src/main/java/hudson/model/StringParameterDefinition.java]: Note that you really need to do `StringParameterValue` too for Pipeline support, though this could be subsumed under the proposed change mentioned in [JENKINS-26093](https://issues.jenkins-ci.org/browse/JENKINS-26093): making that an `Extension` too. [reference](https://github.com/jenkinsci/symbol-plugin/blob/66fdafc0ca3bf733819819479f75fb05f22ae0dc/plugin/src/main/java/org/jenkinsci/plugins/structs/describable/DescribableModel.java#L374-L388)

**Comment by jtnord** [comment]
@jglick the confluence macro is correct.  the plugin itself is inhereting from 
    `org.jenkins-ci.plugins:plugin:1.580.1` which lists the [SCM information](https://github.com/jenkinsci/jenkins/blob/jenkins-1.580.1/plugins/pom.xml#L21-L26) as SVN.
So the plugin is borked - you will be unable to release it independetly from the top level aggregator and as such the annotations versions are currently 1:1 version tied with the plugin implementation.

File: core/src/main/java/hudson/node_monitors/ClockMonitor.java (hunk 6 above)
Comment (jglick) [inline core/src/main/java/hudson/node_monitors/ClockMonitor.java]: Yeah I am not sure I see how this was required now.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 15]
```diff
@@ -784,7 +785,7 @@ public static Jenkins getInstance() {
     private transient final boolean oldJenkinsJVM;
 
     protected Jenkins(File root, ServletContext context) throws IOException, InterruptedException, ReactorException {
-        this(root,context,null);
+        this(root, context, null);
```
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: Please avoid junk hunks.

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 1 above)
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: No, this is for configuring a `views` property from a system DSL script I suppose.

**Comment by jglick** [comment]
> there is no way of updating this library without updating jenkins core

True, but seeing as the library consists of a single annotation with a single attribute, I am not too concerned.

**Comment by jglick** [comment]
> the plugin itself is inheriting from `org.jenkins-ci.plugins:plugin:1.580.1`

I see. I did ask in https://github.com/jenkinsci/structs-plugin/commit/ba025471f4231eafc918597fb1b6a4b86bde523e and https://github.com/jenkinsci/structs-plugin/commit/90c0a8e3da9986fca2a200f7ec7cbf4ff6ec8b23 to fix up the POM.

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 1 above)
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: then :bug: as it requires at least some syncronization as at the moment you can get a junk result.

**Comment by jglick** [comment]
> annotations are designed to fail safe if the annotation is not on the classpath.

Yes, but that does not help here.

> The way of accessing these annotations as I understand it is via the plugin - so if the plugin provided this library then it would be available to it and all other plugins that depended on it

The annotation type would be accessible from the plugin, but IIUC `SomeCoreClass.class.getAnnotation(Symbol.class)` would be null.

**Comment by stephenc** [comment]
> annotations are designed to fail safe if the annotation is not on the classpath.

But if the annotations are not in the classpath **at the time of loading** then the loaded class will be equivalent to a class without the annotation....

So plugins will only see annotations on classes loaded after they were added to the über classloader...

And you still have to deal with two plugins pulling in the different versions of the library.

It's a bit like the fun with bouncycastle only not quite as bad as at least we have control over the annotation evolution

**Comment by jglick** [comment]
> on classes loaded after they were added to the über classloader

There are no classes defined in the `UberClassLoader`, certainly not core classes.

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 1 above)
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: I doubt synchronization is relevant here, since this would only be called from inside a system configuration script, probably during startup. But perhaps this is intended to resynch a Jenkins instance which is already up and running? I do not know.

File: core/src/main/java/hudson/model/TextParameterDefinition.java (hunk 5 above)
Comment (kohsuke) [inline core/src/main/java/hudson/model/TextParameterDefinition.java]: This is because Job DSL plugins uses "xyzParam" names, but it's against what the naming guideline says --- it adds a suffix that refers to the extension point.

So I'm giving the primary name that matches the naming guideline `text` and giving it an alias that matches what Job DSL plugin uses.

File: core/src/main/java/hudson/node_monitors/ClockMonitor.java (hunk 6 above)
Comment (kohsuke) [inline core/src/main/java/hudson/node_monitors/ClockMonitor.java]: Symbol names are given to the type, so for those extensions to get symbols descriptors need to be named type.

File: core/src/main/java/hudson/tasks/BatchFile.java (hunk 11 above)
Comment (kohsuke) [inline core/src/main/java/hudson/tasks/BatchFile.java]: Coming from Job DSL

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 1 above)
Comment (kohsuke) [inline core/src/main/java/jenkins/model/Jenkins.java]: You are right that this is for system config DSL. `@DataBoundSetter` was unintended.

There's no meaningful synchronization that can be done here on `CopyOnWriteArrayList`, and I don't think it's a problem.

File: core/src/main/java/hudson/model/ListView.java (hunk 3 above)
Comment (kohsuke) [inline core/src/main/java/hudson/model/ListView.java]: Correct

File: core/src/main/java/hudson/model/Slave.java (hunk 4 above)
Comment (kohsuke) [inline core/src/main/java/hudson/model/Slave.java]: All the other constructors on this abstract class is public, so I think there's a valid argument to be made both ways. I think it's fine as is.

**Comment by daniel-beck** [comment]
@kohsuke Is there a particular reason this needs to make it into 2.0? Or can we hold off until we've returned to business as usual?

File: core/src/main/java/hudson/model/TextParameterDefinition.java (hunk 5 above)
Comment (jglick) [inline core/src/main/java/hudson/model/TextParameterDefinition.java]: OK…but we should have some sort of consistency I suppose (see `BooleanParameterDefinition`).

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 16]
```diff
@@ -1593,6 +1594,21 @@ public void addView(View v) throws IOException {
         viewGroupMixIn.addView(v);
     }
 
+    /**
+     * Completely replaces views
+     */
+    public void setViews(Collection<View> views) throws IOException {
```
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: Wait, why did you remove the `@DataBoundSetter` here? You need that for `DescribableModel` to recognize `views` as a mutable property.

File: core/src/main/java/hudson/AboutJenkins.java (hunk 12 above)
Comment (kohsuke) [inline core/src/main/java/hudson/AboutJenkins.java]: pipeline plugin only cares about `Descriptors`, but system-config-dsl plugin cares more than that. I think it makes sense to make one sweep to put them everywhere than finding out later that those annotations are missing on things one needs, especially given the long cycle time it takes for a change to get into the core and become usable in, say, LTS.

Comment (kohsuke) [inline core/src/main/java/hudson/AboutJenkins.java]: As an illustration of the use case, I want the system-config-dsl plugin or something similar to be able to filter out extensions for OEM use cases.

File: core/src/main/java/hudson/model/BooleanParameterDefinition.java (hunk 13 above)
Comment (kohsuke) [inline core/src/main/java/hudson/model/BooleanParameterDefinition.java]: See above for my comment about why it has `param` suffix; short answer, job DSL.

The `@Symbol` annotation javadoc requires this to be a valid java identifier, which boolean is not, which is why `boolean` is not here. Do you think that constraint should be relaxed?

File: core/src/main/java/hudson/model/StringParameterDefinition.java (hunk 14 above)
Comment (kohsuke) [inline core/src/main/java/hudson/model/StringParameterDefinition.java]: I'm afraid I don't understand this.

**Comment by kohsuke** [comment]
> Why are you even proposing this for the 2.0 branch? I see no reason why this should not go into master

Jenkins 1.x is dead to me!

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 16 above)
Comment (kohsuke) [inline core/src/main/java/jenkins/model/Jenkins.java]: `Jenkins` is not `Describable` so it shouldn't have `@DataBoundSetter`.

**Comment by kohsuke** [comment]
> @kohsuke Is there a particular reason this needs to make it into 2.0? Or can we hold off until we've returned to business as usual?

As I wrote in the beginning I respect your authority here to postpone it. No problem. I'm putting this up for PR primarily because this change is ready.

The reason I want to have this change in is because that'd let me say system config DSL feature is available in Jenkins 2, and while that was [called out of scope](https://wiki.jenkins-ci.org/display/JENKINS/Jenkins+2.0), it's still [a very popular feature](https://issues.jenkins-ci.org/browse/JENKINS-31094) validated through a number of conferences.

File: core/src/main/java/hudson/AboutJenkins.java (hunk 12 above)
Comment (jglick) [inline core/src/main/java/hudson/AboutJenkins.java]: Seems like a fairly advanced use case to me. Remember that even without `@Symbol` annotations, a DSL is able to fall back to simple or fully-qualified class names. (`SymbolLookup.find` does not currently support this, but should; `DescribableModel.resolveClass` does, though on the other hand it does not currently accept `@Symbol`!)

File: core/src/main/java/hudson/model/BooleanParameterDefinition.java (hunk 13 above)
Comment (jglick) [inline core/src/main/java/hudson/model/BooleanParameterDefinition.java]: > Do you think that constraint should be relaxed?

Not sure. Probably at least one alias should be a valid Java identifier.

File: core/src/main/java/hudson/model/ChoiceParameterDefinition.java [hunk 17]
```diff
@@ -91,7 +92,7 @@ public StringParameterValue createValue(String value) {
         return checkValue(new StringParameterValue(getName(), value, getDescription()));
     }
 
-    @Extension
+    @Extension @Symbol("choiceParam")
```
Comment (jglick) [inline core/src/main/java/hudson/model/ChoiceParameterDefinition.java]: But then why no `choice` alias here?

File: core/src/main/java/hudson/model/StringParameterDefinition.java (hunk 14 above)
Comment (jglick) [inline core/src/main/java/hudson/model/StringParameterDefinition.java]: If the linked code were deleted, we would be unable to support

```groovy
build job: '…', parameters: [[$class: 'StringParameterValue', name: '…', value: '…']]
```

without FQNs, nor could this field be mapped to the syntax proposed in JENKINS-29922.

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 16 above)
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: So then what is `setViews` for?

**Comment by jglick** [comment]
`NodePropertyTest` failures may be real.

**Comment by jglick** [comment]
:bee: (though it would be preferable to see some proposed use case as a linked PR)

**Comment by daniel-beck** [comment]
@kohsuke 

> Jenkins 1.x is dead to me!

Aren't we going to merge `2.0` branch into `master` soon enough? AFAIU any PR opened based on this branch targets the 2.0 release specifically, rather than the 2.x release line.

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 16 above)
Comment (kohsuke) [inline core/src/main/java/jenkins/model/Jenkins.java]: As you noted somewhere else this is so that I can set the exact list of views from system-config-dsl plugin.

It's just that the mechanism I handle `Jenkins` is different from how I handle `Describable`s. The former already exists and I fiddle with its setters, the latter I use `DescribableModel` in the structs plugin.

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 15 above)
Comment (kohsuke) [inline core/src/main/java/jenkins/model/Jenkins.java]: I still haven't figured out why my IDE had started doing it.

File: core/src/main/java/hudson/model/StringParameterDefinition.java (hunk 14 above)
Comment (kohsuke) [inline core/src/main/java/hudson/model/StringParameterDefinition.java]: Ah, so there are classes with `DataBoundConstructor`s that are not `Describable`. I missed that. This needs some more thoughts. So far `@Symbol` was paired up with `@Extension` (which means it's on `Descriptor` and not on `Describable`).

I wonder if maybe we should turn `ParameterValue` into a `Describable`. Or are there others like this?

**Comment by kohsuke** [comment]
Remaining work based on the conversation so far:

* [ ] look at a test failure
* [ ] add `choice` in addition to `choiceParam`
* [ ] structs plugin pom fix
* [ ] think about `@DataBoundConstructor` without being `Describable`

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 18]
```diff
@@ -1593,6 +1594,21 @@ public void addView(View v) throws IOException {
...
+        BulkChange bc = new BulkChange(this);
+        try {
+            this.views.clear();
+            for (View v : views) {
+                addView(v);
```
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: this is still not thread safe.

If 2 threads are calling this you will end up with a junk result that matches neither of the requests.

This also seems to allow you to bypass the check that you can not remove the default view - there is no check that any of the provided views would be set as a default.  (I may be missing some subtlety here).

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 1 above)
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: I would hope at least `addAll(Collection)` minimizes the risk.

File: core/src/main/java/hudson/model/StringParameterDefinition.java (hunk 14 above)
Comment (jglick) [inline core/src/main/java/hudson/model/StringParameterDefinition.java]: > So far `@Symbol` was paired up with `@Extension` (which means it's on `Descriptor` and not on `Describable`).

Hmm, this clashes with my original understanding of the design and the way that `DescribableModel` works, which is at the `Describable` level. Does need more thought.

**Comment by daniel-beck** [comment]
This is not going into 2.0, and for 2.1+ it needs to be re-filed against the master branch.

**Comment by daspilker** [comment]
Please ping me I can help with Job DSL related issues. I'm fine with adding aliases or renaming methods in Job DSL if it improves the DSL. I do not particularly like the `xxxParam` naming for parameters, but it somehow solves the issue for `boolean`.

I want to test the `Symbol` annotation for Job DSL with the Gerrit Trigger plugin as it has a more complex config UI and the support in Job DSL core for Gerrit Trigger needs an overhaul anyway.

**Comment by daniel-beck** [comment]
@kohsuke Please resubmit towards the `master` branch after it has been synced with `2.0` towards 2.1.
